bluemira/geometry/coordinates.py

```diff
@@ -0,0 +1,943 @@
+    z: np.array
+        The z coordinates of the polygon
+
+    Returns
+    -------
+    ccw: bool
+        True if polygon counterclockwise
+    """
+    a = 0
+    for n in range(len(x) - 1):
+        a += (x[n + 1] - x[n]) * (z[n + 1] + z[n])
+    return a < 0
+
+
+def check_ccw_3d(x, y, z, normal):
+    """
+    Check if a set of coordinates is counter-clockwise w.r.t a normal vector.
+
+    Parameters
+    ----------
+    x: np.array
+        The first set of coordinates [m]
+    y: np.array
+        The second set of coordinates [m]
+    z: np.array
+        The third set of coordinates [m]
+    normal: np.array
+        The normal vector about which to check for CCW
+
+    Returns
+    -------
+    ccw: bool
+        Whether or not the set is CCW about the normal vector
+    """
+    r = rotation_matrix_v1v2([0, 0, 1], normal)
+    x, y, z = r @ np.array([x, y, z])
+    a = _get_ccw_metric(x, y, z)
+    return np.dot(normal, a) >= 0.0
+
+
+@nb.jit(cache=True, nopython=True)
+def _get_ccw_metric(x, y, z):
+    """
+    Calculate the signed area of a set of x, y, z coordinate vectors.
+    `Link Shoelace method <https://en.wikipedia.org/wiki/Shoelace_formula>`_
+    """
+    m = np.zeros((3, len(x)))
+    # Translate all coordinates arbitrarily to dodge degenerate edge cases! :)
+    m[0, :] = x + 1.0
+    m[1, :] = y + 1.0
+    m[2, :] = z + 1.0
+    a = np.array([0.0, 0.0, 0.0])
+    for i in range(len(z)):
+        a += np.cross(m[:, i], m[:, (i + 1) % len(z)])
+    return a
+
+
+@xyz_process
+def get_centroid(x, y, z=None):
+    """
+    Calculate the centroid of a non-self-intersecting 2-D counter-clockwise polygon.
+
+    Parameters
+    ----------
+    x: np.array
+        x coordinates of the loop to calculate on
+    y: np.array
+        y coordinates of the loop to calculate on
+    z: Union[None, np.array]
+
+    Returns
+    -------
+    centroid: np.array
+        The x, y, [z] coordinates of the centroid [m]
+    """
+    if z is None:
+        return get_centroid_2d(x, y)
+    else:
+        return get_centroid_3d(x, y, z)
+
+
+@nb.jit(cache=True, nopython=True)
+def get_centroid_2d(x, z):
+    """
+    Calculate the centroid of a non-self-intersecting 2-D counter-clockwise polygon.
+
+    Parameters
+    ----------
+    x: np.array
+        x coordinates of the loop to calculate on
+    z: np.array
+        z coordinates of the loop to calculate on
+
+    Returns
+    -------
+    centroid: List[float]
+        The x, z coordinates of the centroid [m]
+    """
+    if not check_ccw(x, z):
+        x = x[::-1]
+        z = z[::-1]
+    area = get_area_2d(x, z)
+
+    cx, cz = 0, 0
+    for i in range(len(x) - 1):
+        a = x[i] * z[i + 1] - x[i + 1] * z[i]
+        cx += (x[i] + x[i + 1]) * a
+        cz += (z[i] + z[i + 1]) * a
+
+    if area != 0:
+        # Zero division protection
+        cx /= 6 * area
+        cz /= 6 * area
+
+    return [cx, cz]
+
+
+def get_centroid_3d(x, y, z):
```

Comment:
SonarCloud identified this as a duplicate of [`get_centroid_3d` ](https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/9e94a3f4e58d73bab1521a0b5c6468b0333de139/BLUEPRINT/geometry/geomtools.py#L516). Can we remove that BLUEPRINT version and replace it with this one?